### PR TITLE
[path_provider] now supports getDownloadsDirectory() on macOS.

### PR DESCRIPTION
**Pull request by technolion**
## Description

This PR adds support for retrieving a user's downloads directory on macOS using the path_provider plugin. The native code is done in the sub project `path_provider_macos`.

## Related Issues

Fixes flutter/flutter#47676 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by technolion** [comment]
@googlebot I signed it!

**Comment by franciscojma86** [comment]
I just created https://github.com/flutter/flutter/issues/47851 (couldn't tag you for some reason). We need to federate this plugin as there are many platform specific APIs. If you are interested in this, please feel free to drive it :)

For this PR, it might be OK to just add this one API for now since it already most of the current APIs are platform specific. @amirh to make the call here.

File: packages/path_provider/path_provider_macos/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 0.0.3+1
```
Comment (franciscojma86) [inline packages/path_provider/path_provider_macos/CHANGELOG.md]: This should be 0.0.3. We only add +x when it's a change that doesn't affect the API or actual code (readme updates, comments etc)

File: packages/path_provider/path_provider_macos/pubspec.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 name: path_provider_macos
 description: macOS implementation of the path_provider plugin
-version: 0.0.2+1
+version: 0.0.2+2
```
Comment (franciscojma86) [inline packages/path_provider/path_provider_macos/pubspec.yaml]: See comment in CHANGELOG.md. Also, they should match :)

**Comment by technolion** [comment]
This is a very simple change set. When can we expect this to get reviewed and hopefully merged?

File: packages/path_provider/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+## 1.5.2
+
+* macOS support now allows for retrieving the user's downloads directory.
```
Comment (amirh) [inline packages/path_provider/CHANGELOG.md]: nit: I woudn't mention macOS here, from this package's perspective a `getDownloadsDirectory` was added (which is not supported on Android and iOS)

File: packages/path_provider/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,7 @@
+## 1.5.2
```
Comment (amirh) [inline packages/path_provider/CHANGELOG.md]: This is a new feature so should be a minor version bump (1.6.0)

File: packages/path_provider/pubspec.yaml [hunk 5]
```diff
@@ -2,7 +2,7 @@ name: path_provider
 description: Flutter plugin for getting commonly used locations on the Android &
   iOS file systems, such as the temp and app data directories.
 homepage: https://github.com/flutter/plugins/tree/master/packages/path_provider
-version: 1.5.1
+version: 1.5.2
```
Comment (amirh) [inline packages/path_provider/pubspec.yaml]: ditto: 1.6.0

**Comment by technolion** [comment]
Thanks for merging! When is going to be published on [https://pub.dev/packages/path_provider](https://pub.dev/packages/path_provider)

**Comment by technolion** [comment]
> Thanks for merging! When is going to be published on https://pub.dev/packages/path_provider ?
Ah, I see it has been published now. Thanks again!

**Comment by Faiizii** [comment]
` Unhandled Exception: Unsupported operation: Functionality only available on macOS` on calling `var dirs = (await getDownloadsDirectory());`
still facing this issue. here is my `flutter --version` logs

`Flutter 1.23.0-18.1.pre • channel beta • https://github.com/flutter/flutter.git
Framework • revision 198df796aa (5 weeks ago) • 2020-10-15 12:04:33 -0700
Engine • revision 1d12d82d9c
Tools • Dart 2.11.0 (build 2.11.0-213.1.beta)`

let me know if I am doing anything wrong
Note: I already checked for platform before calling `getDownloadsDirectory()`

**Comment by stuartmorgan** [comment]
> channel beta

If you are on `beta`, you can't be running a macOS application, so this issue would not be relevant.

**Comment by technolion** [comment]
Yes, try to 

```
flutter channel dev
flutter upgrade
flutter pub get
```

Then it should work.

**Comment by stuartmorgan** [comment]
@technolion I think you misunderstood my comment. @Faiizii is seeing an error message, and is thus running something. But it's on `beta` channel, which means that whatever that something is, it's not a macOS application. Running that non-macOS Flutter app on a different channel isn't going to change anything about that error, or make this PR relevant.

**Comment by technolion** [comment]
You are right @stuartmorgan. I got confused about by the error message talking about macOS.
The problem was fixed by me in this PR, which removes the pre-screening:
https://github.com/flutter/plugins/pull/3017
Unfortunately this PR was not yet merged. Is there something that is expected from me to fullfill any precondition to get it merged?

**Comment by stuartmorgan** [comment]
> The problem was fixed by me in this PR, which removes the pre-screening

Unless @Faiizii is using a third-party method-channel-based unendorsed implementation of the plugin, that PR will just replace one error method with a different one. There are no implementations in flutter/plugins that are currently gated behind that check.

> Unfortunately this PR was not yet merged. Is there something that is expected from me to fullfill any precondition to get it merged?

As a starting point, it would need to have the test failures addressed. Then it needs a review; there's a large backlog and that PR is low priority since it doesn't fix any current issue (although it could avoid issues in the future, and is the right change to make).

**Pull request merged**